Prints server info on istioctl version

jwendell opened the pull request:
Similar to what `kubectl version` does, this patch adds the ability
to show server's info on the output.

The default behavior is to just show client info. The flag `--remote`
was added to show server information. Fetching server info might take
some time, so, it's an opt-in feature rather than a default behavior.

This patch also modifies the output to match the `kubectl` behavior, i.e.:
 - Dumps the version structure
 - Adds JSON and YAML output formats

This is useful for debugging purposes, where one can easily find
the version of each Istio component in the control plane that's
actually deployed.

Examples of usage:
 - mixs version -o yaml
 - galley version -o json
 - istioctl version --remote
 - istioctl version --remote --short
 - istioctl version --remote -o yaml

Closes #2491.

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **nmittler**

If they are not already assigned, you can assign the PR to them by writing `/assign @nmittler` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[istioctl/OWNERS](https://github.com/istio/istio/blob/master/istioctl/OWNERS)**
- **[pkg/OWNERS](https://github.com/istio/istio/blob/master/pkg/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["nmittler"]} -->

rshriram commented on istioctl/pkg/kubernetes/client.go:
```diff
@@ -217,3 +220,62 @@ func (client *Client) GetPilotAgentContainer(podName, podNamespace string) (stri
+	labelToPodDetail := map[string]podDetail{
+		"pilot":            {"/usr/local/bin/pilot-discovery", "discovery"},
+		"citadel":          {"/usr/local/bin/istio_ca", "citadel"},
+		"egressgateway":    {"/usr/local/bin/pilot-agent", "egressgateway"},
```
this is changing to istio-proxy

rshriram commented on istioctl/pkg/kubernetes/client.go:
```diff
@@ -217,3 +220,62 @@ func (client *Client) GetPilotAgentContainer(podName, podNamespace string) (stri
+
+		res += fmt.Sprintf("Component: %s\n", component)
+		if detail, ok := labelToPodDetail[component]; ok {
+			cmd := []string{detail.binary, "version"}
```
are you sure all of these have the version command?

jwendell replied:
yep

rshriram commented on istioctl/pkg/kubernetes/client.go:
```diff
@@ -217,3 +220,62 @@ func (client *Client) GetPilotAgentContainer(podName, podNamespace string) (stri
+		"citadel":          {"/usr/local/bin/istio_ca", "citadel"},
+		"egressgateway":    {"/usr/local/bin/pilot-agent", "egressgateway"},
+		"galley":           {"/usr/local/bin/galley", "validator"},
+		"ingressgateway":   {"/usr/local/bin/pilot-agent", "ingressgateway"},
```
I don't think we should include gateways. People are likely to deploy their own..

jwendell replied:
The filter I'm doing is getting all pods with `istio` label. If one doesn't deploy a gateway, it won't appear in the list.

istio-testing commented:
@jwendell: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 18b9b8850fce58ebcdadb9f09ba3cc677967a7f8 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7760/istio-presubmit/15553/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

jwendell commented on istioctl/cmd/istioctl/version.go:
```diff
@@ -0,0 +1,67 @@
+				fmt.Fprintln(cmd.OutOrStdout(), pkg_version.Info.LongForm())
+			}
+			if !clientOnly {
+				fmt.Fprintln(cmd.OutOrStdout(), "\n*** Istio Control plane version ***")
```
@liamawhite I thought the same, although I wasn't sure whether to do this in this PR or in another one, to make review easier. What do you think?

jwendell commented:
@liamawhite I've changed the behavior to match the `kubectl` one. Updated the commit message accordingly. Next natural step is to improve the gitVersion string, but it's better to do it in another PR so that this one doesn't become giant.

liamawhite commented:
@jwendell could you provide an example of the new command in your initial comment?

You may also need to make a trivial change in order to retrigger the circleci tests.

istio-testing commented:
@jwendell: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | bcc615927bf61e1cbc73d35561860eac7f06e902 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7760/istio-unit-tests/11636/) | `/test istio-unit-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@jwendell: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | bcc615927bf61e1cbc73d35561860eac7f06e902 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7760/istio-unit-tests/11636/) | `/test istio-unit-tests`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | bcc615927bf61e1cbc73d35561860eac7f06e902 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7760/istio-pilot-e2e-envoyv2-v1alpha3/5541/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@jwendell: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | bcc615927bf61e1cbc73d35561860eac7f06e902 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7760/istio-unit-tests/11636/) | `/test istio-unit-tests`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | bcc615927bf61e1cbc73d35561860eac7f06e902 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7760/istio-pilot-e2e-envoyv2-v1alpha3/5541/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`
prow/istio-pilot-multicluster-e2e.sh | bcc615927bf61e1cbc73d35561860eac7f06e902 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7760/istio-pilot-multicluster-e2e/527/) | `/test istio-pilot-multicluster-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@jwendell: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-bookInfoTests-v1alpha3.sh | 18bbe8e48563aed0fd38cbd5b25a653904d3667c | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7760/e2e-bookInfoTests-envoyv2-v1alpha3/4742/) | `/test e2e-bookInfo-envoyv2-v1alpha3`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

jwendell commented:
/retest

istio-testing commented:
@jwendell: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-bookInfoTests-v1alpha3.sh | 18bbe8e48563aed0fd38cbd5b25a653904d3667c | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7760/e2e-bookInfoTests-envoyv2-v1alpha3/4742/) | `/test e2e-bookInfo-envoyv2-v1alpha3`
prow/istio-unit-tests.sh | 18bbe8e48563aed0fd38cbd5b25a653904d3667c | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/7760/istio-unit-tests/11727/) | `/test istio-unit-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

jwendell commented on istioctl/pkg/kubernetes/client.go:
```diff
@@ -217,3 +223,84 @@ func (client *Client) GetPilotAgentContainer(podName, podNamespace string) (stri
+			var err error
+			var info version.BuildInfo
+
+			stdout, stderr, err = client.PodExec(pod.Name, pod.Namespace, detail.container, cmdJSON)
```
@liamawhite done. Although I replaced just one call. The other one I couldn't because I'm relying on the `stderr` in order to know whether we are dealing with old server components - which don't recognize the `--output` flag. That way we are backward compatible, meaning newer `istioctl` can get version info from old server components.

jwendell commented:
Here are some outputs with this new behavior:
```sh
$ istioctl version 
version.BuildInfo{Version:"2e0baef4078274315464da8bbee1fb8d533ce93f-dirty", GitRevision:"2e0baef4078274315464da8bbee1fb8d533ce93f-dirty", User:"jwendell", Host:"jw-laptop", GolangVersion:"go1.10.3", DockerHub:"docker.io/istio", BuildStatus:"Modified"}

$ istioctl version --short
2e0baef4078274315464da8bbee1fb8d533ce93f-dirty

$ istioctl version -o json
{
  "clientVersion": {
    "version": "2e0baef4078274315464da8bbee1fb8d533ce93f-dirty",
    "revision": "2e0baef4078274315464da8bbee1fb8d533ce93f-dirty",
    "user": "jwendell",
    "host": "jw-laptop",
    "golang_version": "go1.10.3",
    "hub": "docker.io/istio",
    "status": "Modified"
  }
}

$ istioctl version --remote
client version: version.BuildInfo{Version:"2e0baef4078274315464da8bbee1fb8d533ce93f-dirty", GitRevision:"2e0baef4078274315464da8bbee1fb8d533ce93f-dirty", User:"jwendell", Host:"jw-laptop", GolangVersion:"go1.10.3", DockerHub:"docker.io/istio", BuildStatus:"Modified"}
citadel version: version.BuildInfo{Version:"1.0.0", GitRevision:"3a136c90ec5e308f236e0d7ebb5c4c5e405217f4", User:"root@71a9470ea93c", Host:"", GolangVersion:"go1.10.1", DockerHub:"docker.io/istio", BuildStatus:"Clean"}
egressgateway version: version.BuildInfo{Version:"1.0.0", GitRevision:"3a136c90ec5e308f236e0d7ebb5c4c5e405217f4", User:"root@71a9470ea93c", Host:"", GolangVersion:"go1.10.1", DockerHub:"docker.io/istio", BuildStatus:"Clean"}
galley version: version.BuildInfo{Version:"1.0.0", GitRevision:"3a136c90ec5e308f236e0d7ebb5c4c5e405217f4", User:"root@71a9470ea93c", Host:"", GolangVersion:"go1.10.1", DockerHub:"docker.io/istio", BuildStatus:"Clean"}
ingressgateway version: version.BuildInfo{Version:"1.0.0", GitRevision:"3a136c90ec5e308f236e0d7ebb5c4c5e405217f4", User:"root@71a9470ea93c", Host:"", GolangVersion:"go1.10.1", DockerHub:"docker.io/istio", BuildStatus:"Clean"}
pilot version: version.BuildInfo{Version:"1.0.0", GitRevision:"3a136c90ec5e308f236e0d7ebb5c4c5e405217f4", User:"root@71a9470ea93c", Host:"", GolangVersion:"go1.10.1", DockerHub:"docker.io/istio", BuildStatus:"Clean"}
policy version: version.BuildInfo{Version:"1.0.0", GitRevision:"3a136c90ec5e308f236e0d7ebb5c4c5e405217f4", User:"root@71a9470ea93c", Host:"", GolangVersion:"go1.10.1", DockerHub:"docker.io/istio", BuildStatus:"Clean"}
sidecar-injector version: version.BuildInfo{Version:"1.0.0", GitRevision:"3a136c90ec5e308f236e0d7ebb5c4c5e405217f4", User:"root@71a9470ea93c", Host:"", GolangVersion:"go1.10.1", DockerHub:"docker.io/istio", BuildStatus:"Clean"}
telemetry version: version.BuildInfo{Version:"1.0.0", GitRevision:"3a136c90ec5e308f236e0d7ebb5c4c5e405217f4", User:"root@71a9470ea93c", Host:"", GolangVersion:"go1.10.1", DockerHub:"docker.io/istio", BuildStatus:"Clean"}

$ istioctl version --remote --short
client version: 2e0baef4078274315464da8bbee1fb8d533ce93f-dirty
citadel version: 1.0.0
egressgateway version: 1.0.0
galley version: 1.0.0
ingressgateway version: 1.0.0
pilot version: 1.0.0
policy version: 1.0.0
sidecar-injector version: 1.0.0
telemetry version: 1.0.0

$ istioctl version --remote -o yaml
clientVersion:
  golang_version: go1.10.3
  host: jw-laptop
  hub: docker.io/istio
  revision: 2e0baef4078274315464da8bbee1fb8d533ce93f-dirty
  status: Modified
  user: jwendell
  version: 2e0baef4078274315464da8bbee1fb8d533ce93f-dirty
meshVersion:
- Component: citadel
  Info:
    golang_version: go1.10.1
    host: ""
    hub: docker.io/istio
    revision: 3a136c90ec5e308f236e0d7ebb5c4c5e405217f4
    status: Clean
    user: root@71a9470ea93c
    version: 1.0.0
- Component: egressgateway
  Info:
    golang_version: go1.10.1
    host: ""
    hub: docker.io/istio
    revision: 3a136c90ec5e308f236e0d7ebb5c4c5e405217f4
    status: Clean
    user: root@71a9470ea93c
    version: 1.0.0
- Component: galley
  Info:
    golang_version: go1.10.1
    host: ""
    hub: docker.io/istio
    revision: 3a136c90ec5e308f236e0d7ebb5c4c5e405217f4
    status: Clean
    user: root@71a9470ea93c
    version: 1.0.0
- Component: ingressgateway
  Info:
    golang_version: go1.10.1
    host: ""
    hub: docker.io/istio
    revision: 3a136c90ec5e308f236e0d7ebb5c4c5e405217f4
    status: Clean
    user: root@71a9470ea93c
    version: 1.0.0
- Component: pilot
  Info:
    golang_version: go1.10.1
    host: ""
    hub: docker.io/istio
    revision: 3a136c90ec5e308f236e0d7ebb5c4c5e405217f4
    status: Clean
    user: root@71a9470ea93c
    version: 1.0.0
- Component: policy
  Info:
    golang_version: go1.10.1
    host: ""
    hub: docker.io/istio
    revision: 3a136c90ec5e308f236e0d7ebb5c4c5e405217f4
    status: Clean
    user: root@71a9470ea93c
    version: 1.0.0
- Component: sidecar-injector
  Info:
    golang_version: go1.10.1
    host: ""
    hub: docker.io/istio
    revision: 3a136c90ec5e308f236e0d7ebb5c4c5e405217f4
    status: Clean
    user: root@71a9470ea93c
    version: 1.0.0
- Component: telemetry
  Info:
    golang_version: go1.10.1
    host: ""
    hub: docker.io/istio
    revision: 3a136c90ec5e308f236e0d7ebb5c4c5e405217f4
    status: Clean
    user: root@71a9470ea93c
    version: 1.0.0
```

sdake commented:
@jwendell looks pretty cool!  I noticed you ran a /retest - if circleci is failing you may consider a rebase to solve the problem.

jwendell commented:
@sdake It's rebased already...

liamawhite commented:
/lgtm

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/7760#issuecomment-417638045" title="LGTM">liamawhite</a>*
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **nmittler**

If they are not already assigned, you can assign the PR to them by writing `/assign @nmittler` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- ~~[istioctl/OWNERS](https://github.com/istio/istio/blob/master/istioctl/OWNERS)~~ [liamawhite]
- **[pkg/OWNERS](https://github.com/istio/istio/blob/master/pkg/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["nmittler"]} -->

jwendell commented:
/assign @geeknoid

ldemailly commented:
/lgtm
/approve

istio-testing commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/7760#issuecomment-420403430" title="Approved">ldemailly</a>*, *<a href="https://github.com/istio/istio/pull/7760#issuecomment-417638045" title="LGTM">liamawhite</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[istioctl/OWNERS](https://github.com/istio/istio/blob/master/istioctl/OWNERS)~~ [ldemailly,liamawhite]
- ~~[pkg/OWNERS](https://github.com/istio/istio/blob/master/pkg/OWNERS)~~ [ldemailly]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

geeknoid commented:
Please add unit tests for all new functionality in the pkg/version package. I think it was at 100% coverage before, so let's keep it there.

Thanks.

geeknoid commented:
/hold

istio-testing commented:
New changes are detected. LGTM label has been removed.

jwendell commented:
@geeknoid done. Now it's 100% covered!

geeknoid commented:
Thanks.

geeknoid commented:
/lgtm
/approve

istio-testing commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/7760#issuecomment-420704033" title="Approved">geeknoid</a>*, *<a href="https://github.com/istio/istio/pull/7760#issuecomment-420403430" title="Approved">ldemailly</a>*, *<a href="https://github.com/istio/istio/pull/7760#issuecomment-417638045" title="LGTM">liamawhite</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[istioctl/OWNERS](https://github.com/istio/istio/blob/master/istioctl/OWNERS)~~ [geeknoid,ldemailly,liamawhite]
- ~~[pkg/OWNERS](https://github.com/istio/istio/blob/master/pkg/OWNERS)~~ [geeknoid,ldemailly]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

jwendell commented:
@geeknoid thanks, should the do-not-merge/hold label be removed?

linsun commented:
nice feature added @jwendell - confirmed code coverage didn't regress.

ldemailly commented:
yes that's done with "/hold cancel" (or removing it form the labels) 
(no need to force merge)
